asab/api/doc.py

```diff
@@ -7,353 +7,400 @@
-						(qp.error_description ? qp.error_description+ ". " : "no accessCode received from the server. ") +
-						(qp.error_uri ? "More info: "+qp.error_uri : "");
-				}
-
-				oauth2.errCb({
-					authId: oauth2.auth.name,
-					source: "auth",
-					level: "error",
-					message: oauthErrorMsg || "[Authorization failed]: no accessCode received from the server."
-				});
-			}
-		} else {
-			oauth2.callback({auth: oauth2.auth, token: qp, isValid: isValid, redirectUrl: redirectUrl});
-		}
-		window.close();
-	}
-
-	if (document.readyState !== 'loading') {
-		run();
-	} else {
-		document.addEventListener('DOMContentLoaded', function () {
-			run();
-		});
-	}
-</script>
-</body>
-</html>"""
-		return aiohttp.web.Response(text=page, content_type="text/html")
-
-
-	async def openapi(self, request):
-		'''
-		Download OpenAPI (version 3) API documentation (aka Swagger) in YAML.
-		---
-		tags: ['asab.doc']
-		externalDocs:
-			description: OpenAPI Specification
-			url: https://swagger.io/specification/
-
-		'''
-		return aiohttp.web.Response(
-			text=yaml.dump(self.build_swagger_specs()),
-			content_type="text/yaml"
-		)
+    def __init__(self, api_service, app, web_container, config_section_name="asab:doc"):
+        self.App = app
+        self.WebContainer = web_container
+        self.WebContainer.WebApp.router.add_get("/doc", self.doc)
+        self.WebContainer.WebApp.router.add_get(
+            "/oauth2-redirect.html", self.oauth2_redirect
+        )
+        self.WebContainer.WebApp.router.add_get("/asab/v1/openapi", self.openapi)
+
+        self.AuthorizationUrl = asab.Config.get(
+            config_section_name, "authorizationUrl", fallback=None
+        )
+        self.TokenUrl = asab.Config.get(config_section_name, "tokenUrl", fallback=None)
+        self.Scopes = asab.Config.get(config_section_name, "scopes", fallback=None)
+
+        self.Manifest = api_service.Manifest
+
+    def build_swagger_documentation(self) -> dict:
+        """Take a docstring of the class and a docstring of methods and merge them into Swagger data."""
+        specification: dict = {}
+        # TODO: check for the add_dict in the primary code
+        app_doc_string: str = self.App.__doc__
+        additional_info_dict: dict = {}
+
+        app_description: str = self.get_description(app_doc_string)
+        additional_info_dict.update(self.get_additional_info(app_doc_string))
+        if additional_info_dict is not None:
+            specification.update(additional_info_dict)
+        
+        specification.update(self.prepare_description_frame(app_description))
+        specification["components"] = self.create_security_schemes()
+        specification["info"]["version"] = self.get_manifest()
+        specification["info"]["description"] = self.get_server_and_container_info(
+            app_description
+        )
+
+        # routers sorting
+        asab_routers = []
+        doc_routers = []
+        microservice_routers = []
```

Comment:
"routes", not "routers"

also, you probably don't need to sort them into three lists. two is just enough. doc_routes are subset of asab_routes, there is no need to differentiate between them, as long as they come after microservice_routes.